### DISCUSSION TITLE
Lapce Thoughts + Wishlist

### DISCUSSION THREAD

**Discussion by cart** [lapce/lapce#626] [General]
I just opened up [Bevy](https://github.com/bevyengine/bevy) in Lapce and I've gotta say I'm both very impressed and excited for the future. I'm a long time VSCode user and Lapce feels like a breath of fresh air:

* Lapce opened and was responsive instantly, compared to a few seconds after opening bevy in vscode
* Its seems like the Rust plugin prepares itself / started autocompleting _much_ faster? Maybe RA was able to reuse state from the VSCode instance or something, but the difference was pretty stark.
* The git integration updates the list of changed files instantly, vs VSCode's 1-3 seconds of delay
* Autocomplete (and typing in general) feels much snappier / less prone to stuttering
* Rust error reporting works as expected, even across files (with one pretty important exception ... see my next list)
* Modal editing already does pretty much everything I need it to 

The smoothness and snappiness makes me want to switch to this as my main driver for Bevy dev _right now_. And most of the things I need are already there. But realistically I can't switch until there are a few more things. Heres my "things i need to switch" list (in rough priority order):

* Global cross-file symbol search. Rust encourages multiple symbols per file (many of which have names like `mod.rs`), so I _need_ global symbol search to efficiently navigate the Bevy code base. This is the one big thing preventing me from switching. VScode has local "per file" and global "per workspace" symbol search modes and i use both regularly. 
* It looks like I don't get error messages until I save a file, and only for that specific file when it is open. I need to be able to see all errors in a workspace, as soon as the workspace opens up. 
* I couldn't find a shortcut to force autocomplete to open up. Sometimes I want to see all symbols available at a given location, before typing a specific letter. I dont want to type every letter of the alphabet to see what my options are.
* I use VSCodes "revert changes to file" option in their git interface a lot. Lapce doesn't seem to have that.  Being able to select multiple files at once and revert them all at once is also really nice.

And here are some smaller things that are nice to have:
* In vim (and vscode's vim emulation), I remap "exit insert mode" to a combination of two letters on my home row (T + N, because I use the colemak layout). This is a pretty common pattern in vim, but it doesn't work for me. Lapce lets me set the binding, but it wont register.
* Entering "insert mode", then exiting "insert mode" results in the cursor shifting to the left by one character. Do this multiple times and you continue shifting to the left. This feels wrong / not a helpful behavior :)
* Markdown syntax highlighting would be nice

**Comment by cart** [comment]
I also love that Lapce is "actually open source" instead of the pseudo-open-source vscode approach. That alone makes me want to switch asap.

**Reply by follower** [reply]
> That alone makes me want to switch asap.

I mean, I don't like to tell anyone what to do, but having said that:

**Do it! :D** 

----

Sure, I'm not using it _daily_ but I did make it through multiple game jams with Lapce & Bevy including at a point in time when...

...using the keyboard shortcut for undo frequently caused a panic...

...and nearly every time I pasted something it got pasted _three_ times unless I was _very_ careful with the key press order...

 ...and if I didn't get it _quite_ right, I could always hit undo, right?! :D

----

Anyway, in case my "vote" makes any difference whatsoever:

 * I'd totally take the delay to Bevy v0.8 if it meant Lapce could lay claim to being _the "Preferred 'Actually Open Source' Editor" of the "Creator of Bevy Engine"_. :)

**Reply by cart** [reply]
Yup I'll be using it off and on as they add features / fix issues, but I can't endorse Lapce as my "preferred editor" until it has a complete and stable Rust workflow. I won't endorse/recommend it "too early", as that would have real consequences for Bevy users.

**Comment by cart** [comment]
The ability to make changes directly in git diffs (with autocomplete) would also be much appreciated.

**Comment by Zooce** [comment]
Wow - Lapce is pretty awesome. I'm genuinely surprised at just how snappy it really is. You got me really excited now!

**Comment by konstantinkopka** [comment]
Why are there so many *awesome* Rust projects I had never heard of, and they're already at 12.4k Github stars 😂
This is a really exciting editor!

**Comment by follower** [comment]
> Its seems like the Rust plugin prepares itself / started autocompleting _much_ faster? Maybe RA was able to reuse state from the VSCode instance or something [...]

FWIW that seems _unlikely_ given the lack of enthusiasm for caching state to help a *single* editor instance: https://github.com/rust-lang/rust-analyzer/issues/4712#issuecomment-637758082

----

\*laugh\* \*cry\* _while trying to write Bevy jam games with Lapce on a..._

..."Dual-Core Intel® Core™ i7-4510U CPU @ 2.00GHz" with "Intel Corporation Haswell-ULT Integrated Graphics Controller (rev 0b)"...

...with an iGPU driver that figures it'll free its memory when the OOM killer approaches--good luck fighting it out with Firefox...

...with 8GB RAM and an ADHD-driven FF window/tabs collection in the...very many.

(To be fair, Mozilla have *massively* improved the practicality of this impracticality over the past year or so.)

Okay, realistically, I _did_ pretty much give up on trying to make that combination work and do most of my Bevy game dev with the _slightly_ more practical...

..."Dual-Core Intel® Celeron® CPU G530 @ 2.40GHz" with "NVIDIA Corporation GK106 [GeForce GTX 650 Ti]" 8GB RAM...

...that currently isn't waking from suspend properly. :D \*sigh\*

----

Wait, what was I saying?

**Oh, yeah:**

 * **Lapce & Bevy: two great tastes that go great together, with thanks to y'all [from RancidBacon and Limited Attention Studios](https://rancidbacon.itch.io)! :)**

(Okay, _sure_, Godot is still ahead there on sheer numbers--but, you know, Bevy doesn't even _have_ an _editor_!? Shout out to the `egui`, `bevy_egui` & `bevy_inspector_egui` crew--the latter of which I _finally_ started using recently, long after it would have been wise to start. :D )

**Reply by cart** [reply]
I think I've sorted out why startup is faster. VSCode has two init phases: "RA indexing" and "cargo check for the whole workspace". Lapce only does the "RA indexing" phase at startup. I suspect this is why we don't see "global error messages" in lapce.
